lib/std/os/linux/tls.zig

```diff
@@ -195,124 +192,138 @@ pub fn initTLS() ?*elf.Phdr {
+    };
 }
 
-pub fn copyTLS(addr: usize) usize {
-    const tls_img = tls_image.?;
+inline fn alignPtrCast(comptime T: type, ptr: [*]u8) *T {
+    return @ptrCast(*T, @alignCast(@alignOf(*T), ptr));
+}
 
-    // Be paranoid, clear the area we're going to use
-    @memset(@intToPtr([*]u8, addr), 0, tls_img.alloc_size);
+/// Initializes all the fields of the static TLS area and returns the computed
+/// architecture-specific value of the thread-pointer register
+pub fn prepareTLS(area: []u8) usize {
+    // Clear the area we're going to use, just to be safe
+    mem.set(u8, area, 0);
     // Prepare the DTV
-    const dtv = @intToPtr(*DTV, addr + tls_img.dtv_offset);
+    const dtv = alignPtrCast(DTV, area.ptr + tls_image.dtv_offset);
     dtv.entries = 1;
-    dtv.tls_block[0] = addr + tls_img.data_offset + tls_dtv_offset;
-    // Set-up the TCB
-    // Force the alignment to 1 byte as the TCB may start from a non-aligned
-    // address under the variant II model
-    const tcb_ptr = @intToPtr(*align(1) usize, addr + tls_img.tcb_offset);
-    if (tls_variant == TLSVariant.VariantI) {
-        tcb_ptr.* = addr + tls_img.dtv_offset;
-    } else {
-        tcb_ptr.* = addr + tls_img.tcb_offset;
-    }
+    dtv.tls_block[0] = area.ptr + tls_dtv_offset + tls_image.data_offset;
+    // Prepare the TCB
+    const tcb_ptr = alignPtrCast([*]u8, area.ptr + tls_image.tcb_offset);
+    tcb_ptr.* = switch (tls_variant) {
+        .VariantI => area.ptr + tls_image.dtv_offset,
+        .VariantII => area.ptr + tls_image.tcb_offset,
+    };
     // Copy the data
-    @memcpy(@intToPtr([*]u8, addr + tls_img.data_offset), tls_img.data_src.ptr, tls_img.data_src.len);
+    mem.copy(u8, area[tls_image.data_offset..], tls_image.data_src);
 
     // Return the corrected (if needed) value for the tp register
-    return addr + tls_tp_offset +
-        if (tls_tp_points_past_tcb) tls_img.data_offset else tls_img.tcb_offset;
+    return @ptrToInt(area.ptr) + tls_tp_offset +
+        if (tls_tp_points_past_tcb) tls_image.data_offset else tls_image.tcb_offset;
 }
 
 var main_thread_tls_buffer: [256]u8 align(32) = undefined;
 
-pub fn allocateTLS(size: usize) usize {
-    // Small TLS allocation, use our local buffer
-    if (size < main_thread_tls_buffer.len) {
-        return @ptrToInt(&main_thread_tls_buffer);
-    }
+pub fn initStaticTLS() void {
+    initTLS();
 
-    const slice = os.mmap(
-        null,
-        size,
-        os.PROT_READ | os.PROT_WRITE,
-        os.MAP_PRIVATE | os.MAP_ANONYMOUS,
-        -1,
-        0,
-    ) catch @panic("out of memory");
+    var tls_area = blk: {
+        // Fast path for the common case where the TLS data is really small,
+        // avoid an allocation and use our local buffer
+        if (tls_image.alloc_size < main_thread_tls_buffer.len) {
+            break :blk main_thread_tls_buffer[0..tls_image.alloc_size];
+        }
 
-    return @ptrToInt(slice.ptr);
+        break :blk os.mmap(
```

Comment:
Should this have a check for `alloc_size > 0`? I don't think we want to do an `mmap` here if no tls is required.

Reply:
The `alloc_size` is never zero.

> don't think we want to do an mmap here if no tls is required

That's the point, we want a per-thread local block of memory even if there's no threadlocal stuff.